### PR TITLE
HHH-15654 Criteria APIs for window and ordered-set aggregate functions

### PR DESCRIPTION
**Pull request by mbladel**
https://hibernate.atlassian.net/browse/HHH-15654

### CODE REVIEW

**Comment by gavinking** [comment]
This is looking really nice. 

One comment: I would much prefer to see overloads of methods like `listagg()` than asking users to pass in null arguments.

**Comment by mbladel** [comment]
> One comment: I would much prefer to see overloads of methods like `listagg()` than asking users to pass in null arguments.

Sure thing, I thought about it but was worried of making the interface "too crowded". Added them now.

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 1]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 * Create a {@code percentile_cont} ordered set-aggregate function expression.
+	 *
+	 * @param order order by clause used in within group
+	 * @param argument argument to the function
+	 *
+	 * @return ordered set-aggregate expression
+	 *
+	 * @see #percentileCont(JpaOrder, JpaPredicate, Expression) percentileCont
+	 */
+	JpaExpression<Integer> percentileCont(JpaOrder order, Expression<? extends Number> argument);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1115]: I think this should rather be:

```suggestion
	<T> JpaExpression<T> percentileCont(Expression<? extends Number> argument, Expression<T> sortExpression, SortOrder sortOrder, NullPrecedence nullPrecedence);
```

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 2]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 *
+	 * @param order order by clause used in within group
+	 * @param filter optional filter clause
+	 * @param argument argument to the function
+	 *
+	 * @return ordered set-aggregate expression
+	 *
+	 * @see #functionWithinGroup
+	 */
+	JpaExpression<Integer> percentileCont(JpaOrder order, JpaPredicate filter, Expression<? extends Number> argument);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1128]: I think this should rather be:

```suggestion
	<T> JpaExpression<T> percentileCont(Expression<? extends Number> argument, JpaPredicate filter, Expression<T> sortExpression, SortOrder sortOrder, NullPrecedence nullPrecedence);
```

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 3]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 *
+	 * @param order order by clause used in within group
+	 * @param filter optional filter clause
+	 * @param argument argument to the function
+	 *
+	 * @return ordered set-aggregate expression
+	 *
+	 * @see #functionWithinGroup
+	 */
+	JpaExpression<Integer> percentileDisc(JpaOrder order, JpaPredicate filter, Expression<? extends Number> argument);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1153]: I think this should rather be:

```suggestion
	<T> JpaExpression<T> percentileDisc(Expression<? extends Number> argument, JpaPredicate filter, Expression<T> sortExpression, SortOrder sortOrder, NullPrecedence nullPrecedence);
```

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 4]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 * Create a {@code percentile_disc} ordered set-aggregate function expression.
+	 *
+	 * @param order order by clause used in within group
+	 * @param argument argument to the function
+	 *
+	 * @return ordered set-aggregate expression
+	 *
+	 * @see #percentileDisc(JpaOrder, JpaPredicate, Expression) percentileDisc
+	 */
+	JpaExpression<Integer> percentileDisc(JpaOrder order, Expression<? extends Number> argument);
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1140]: I think this should rather be:

```suggestion
	<T> JpaExpression<T> percentileDisc(Expression<? extends Number> argument, Expression<T> sortExpression, SortOrder sortOrder, NullPrecedence nullPrecedence);
```

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+					JpaWindow window = cb.createWindow().orderBy( cb.desc( root.get( "theInt" ) ) );
+					JpaExpression<Integer> nthValue = cb.nthValue(
+							root.get( "theInt" ),
+							cb.literal( 2 ),
+							window
+					);
+
+					// todo marco : db2 throws java.lang.IndexOutOfBoundsException: Index 1 out of bounds for length 1
+					//  on getResultList() line, and SqlServer doesn't support nth_value function (even tho it's in Dialect)
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java:175]: Here is a reference for the lack of the function in SQL Server: https://learn.microsoft.com/en-us/openspecs/sql_standards/ms-tsqliso02/16e25dcc-6f2c-4c15-9e71-1d8935c7a6ac

Not sure about DB2, but we can look into the stack trace together if you want.

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+					assertEquals( 26L, resultList.get( 4 ) );
+				}
+		);
+	}
+
+	//	@Test
+	public void testFrame(SessionFactoryScope scope) {
+		scope.inTransaction(
+				session -> {
+					// todo marco : how should we do window function frames in Criteria ?
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java:267]: Just an idea for an API, something like this:

```java
jpaWindow.frameRows( cb.frameBetweenPreceeding( 2 ), cb.frameCurrentRow() );
jpaWindow.frameExclude( FrameExclusion.CURRENT_ROW )
```

with variants `frameRange` and `frameGroups` to control the frame mode. The first argument is the "start" and the second argument the "end" expression of the frame.

This way, the frame is set as a whole (in accordance to the JPA Criteria API style), yet it is still somewhat readable and typesafe. I guess it might be nice to also expose this frame information through the `JpaWindow` contract like e.g.:

```java
interface JpaWindowFrame {
    FrameKind getKind();
    Expression getFrameExpression();
}
```

**Review by beikov — COMMENTED** [review]
Looks like a great first attempt at this. I looked over it like you requested and provided some feedback and some ideas for supporting the rest of the window/aggregate functionalities.

**Comment by mbladel** [comment]
> Looks like a great first attempt at this. I looked over it like you requested and provided some feedback and some ideas for supporting the rest of the window/aggregate functionalities.

Thanks @beikov, I'll go over your suggestions and update the PR

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 7]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 * @param name name of the ordered set-aggregate function
+	 * @param type type of this expression
+	 * @param order order by clause used in within group
+	 * @param filter optional filter clause
+	 * @param args optional arguments to the function
+	 * @param <T> type of this expression
+	 *
+	 * @return ordered set-aggregate function expression
+	 */
+	<T> JpaExpression<T> functionWithinGroup(
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1027]: Yes though about this but wanted to ask before implementing everything.
IIUC at the end we should have 4 overrides for each ordered-set aggregate function:
- `order`
- `order` + `filter`
- `order` + `window`
- `order` + `filter` + `window`

Will also try to add non ordered-set `functionAggregate`s with `filter` and `window` overrides.

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1027]: Yeah, exactly.

File: hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java [hunk 8]
```diff
@@ -892,4 +892,313 @@ <Y extends Comparable<? super Y>> JpaPredicate between(
...
+	 * Create a {@code mode} ordered set-aggregate function expression.
+	 *
+	 * @param order order by clause used in within group
+	 * @param argument argument to the function
+	 *
+	 * @return ordered set-aggregate expression
+	 *
+	 * @see #mode(JpaOrder, JpaPredicate, Expression) mode
+	 */
+	JpaExpression<?> mode(JpaOrder order, Expression<?> argument);
```
[the remark this answers is not in the document]
Comment (mbladel) [reply hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1090]: Should I also substitute the `JpaOrder` parameter with `SortOrder` + `NullPrecedence` for the generic `functionWithinGroup` and `listagg` then?

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/query/criteria/HibernateCriteriaBuilder.java:1090]: No, this is just for these few functions where the single order by item dictates the result type.

File: hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java (hunk 6 above)
Comment (beikov) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java]: Great work. I think this is as nice as this can get. Any opinions @sebersole @dreab8 ?

Comment (mbladel) [reply hibernate-core/src/test/java/org/hibernate/orm/test/query/criteria/CriteriaWindowFunctionTest.java]: TY so much @beikov @dreab8, couldn't have done it without your help :smile:

Just pushing last changes for a couple edge-case errors for some dbs then I should be done.

**Comment by beikov** [comment]
Could you please add `@Incubating` to the new API methods and classes/interfaces? After that, we can merge this.
